File: modules/remotebackend/unittest_zeromq.rb
#!/usr/bin/env ruby

require 'rubygems'
require 'bundler/setup'
require 'json'
require 'zero_mq'
require './unittest'

h = Handler.new()
f = File.open "/tmp/remotebackend.txt.#{$$}","a"
f.sync = true

runcond=true

trap('INT') { runcond = false }
trap('TERM') { runcond = false }

begin
  context = ZeroMQ::Context.new
  socket = context.socket ZMQ::REP
  socket.bind("ipc:///tmp/remotebackend.0")
 
  print "[#{Time.now.to_s}] ZeroMQ unit test responder running\n"

  while(runcond) do
    line = ""
    rc = socket.recv_string line
    # expect json
    input = {}
    line = line.strip

    f.puts "#{Time.now.to_f}: [zmq] #{line}"
    next if line.empty?
    begin
      input = JSON.parse(line)
      method = "do_#{input["method"].downcase}"
      args = input["parameters"] || []

      if h.respond_to?(method.to_sym) == false
         res = false
      elsif args.size > 0
         res, log = h.send(method,args)
      else
         res, log = h.send(method)
      end
      socket.send_string ({:result => res, :log => log}).to_json + "\n" , 0
      f.puts "#{Time.now.to_f} [zmq]: #{({:result => res, :log => log}).to_json}"
    rescue JSON::ParserError
      socket.send_string ({:result => false, :log => "Cannot parse input #{line}"}).to_json + "\n";
      f.puts "#{Time.now.to_f} [zmq]: #{({:result => false, :log => "Cannot parse input #{line}"}).to_json}"
      next
    end
  end
rescue SystemExit, Interrupt, Errno::EINTR
end

print "[#{Time.now.to_s}] ZeroMQ unit test responder ended\n"

File: modules/remotebackend/unittest_zeromq.py
#!/usr/bin/env python

import zmq
import json
import os
from pdns_unittest import Handler


def run(socket, handler):
    while True:
        message = socket.recv()
        try:
            message = json.loads(message.decode().strip())
            method = "do_%s" % message["method"].lower()
            args = message["parameters"]
            handler.result = False
            handler.log = []
            if callable(getattr(handler, method, None)):
                getattr(handler, method)(**args)
                result = json.dumps({"result": handler.result, "log": handler.log})
                socket.send(result.encode())
        except KeyboardInterrupt as e3:
            return
        except BrokenPipeError as e2:
            raise e2
        except Exception as e:
            print(e)
            socket.send(json.dumps({"result": False}).encode())


def main():
    context = zmq.Context()
    socket = context.socket(zmq.REP)
    socket.bind("ipc:///tmp/remotebackend.0")
    handler = Handler()
    print("Listening on ipc:///tmp/remotebackend.0")

    try:
        run(socket, handler)
    except KeyboardInterrupt as e:
        pass

    os.unlink("/tmp/remotebackend.0")


main()
